webapp/src/main/java/com/box/l10n/mojito/smartling/SmartlingClient.java

```diff
@@ -0,0 +1,188 @@
+    }
+
+    Map<String, Object> getPayloadMapRefresh() {
+        Map<String, Object> map = new HashMap<>();
+        map.put("refreshToken", refreshToken);
+        return map;
+    }
+
+    Map<String, Object> getPayloadMapPullSourceString() {
+        return new HashMap<>();
+    }
+
+    HttpHeaders getHttpHeadersForJson() {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        return headers;
+    }
+
+    HttpHeaders getHttpHeadersForJsonWithAuthorization() {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        headers.set("Authorization", "Bearer " + authToken);
+        return headers;
+    }
+
+    <T> HttpEntity<T> getHttpEntityForPayload(T payload) {
+        return new HttpEntity<>(payload, getHttpHeadersForJson());
+    }
+
+    <T> HttpEntity<T> getHttpEntityForPayloadWithAuthorization(T payload) {
+        return new HttpEntity<>(payload, getHttpHeadersForJsonWithAuthorization());
+    }
+
+    public void authenticate() throws SmartlingClientException {
+        Map<String, Object> payload = getPayloadMapAuthentication();
+        AuthenticationResponse authenticationResponse;
+        try {
+            authenticationResponse = postForAuthentication(payload, AUTH_API_V2_AUTHENTICATE);
+        } catch (SmartlingClientException e) {
+            throw new SmartlingClientException(e.getMessage());
+        }
+        if (authenticationResponse.isAuthenticationErrorResponse()) {
+            throw new SmartlingClientException("Error authenticating with Smartling API, please check your credentials");
+        } else if (!authenticationResponse.isSuccessResponse()) {
+            throw new SmartlingClientException(authenticationResponse.getErrorMessage());
+        } else {
+            Instant now = Instant.now();
+            lastAuthentication = now;
+            lastRefresh = now;
+            nextAuthentication = now.plus(Duration.ofHours(authenticationDurationHours));
+            nextRefresh = now.plus(Duration.ofMinutes(refreshDurationMinutes));
+            authToken = authenticationResponse.getResponse().getData().getAccessToken();
+            refreshToken = authenticationResponse.getResponse().getData().getRefreshToken();
+        }
+    }
+
+    AuthenticationResponse postForAuthentication(Map<String, Object> payload, String authApiV2) throws SmartlingClientException {
+        HttpEntity<Map<String, Object>> httpEntity = getHttpEntityForPayload(payload);
+        String postForObject = restTemplate.postForObject(
+                authApiV2, httpEntity, String.class);
+        AuthenticationResponse authenticationResponse;
+        try {
+            authenticationResponse = objectMapper.readValue(postForObject, AuthenticationResponse.class);
+        } catch (IOException e) {
+            logger.error(e.getMessage());
+            throw new SmartlingClientException(e.getMessage());
```

Comment:
In general, it is is better to have an explicit message about what is failing and log/chain the exception.
```
} catch (IOException e) {
            String msg = "Could not deserialize the authentication response";
            logger.error(msg, e); 
            throw new SmartlingClientException(msg, e);
}
```

but in that case you should be able to let the restTemplate do the Jackson deserialization instead of getting a String and then converting with the object mapper.

Reply:
restTemplate doesn't do well with nested classes which is why I convert from string to the class with jackson

Reply:
hm, I'm surprised since it should be using Jackson too and we've been using it without issue. Can you show the exact failure with nested classes? it should probably be fixable.

Reply:
I don't have the logs anymore so I'd have to recreate the conditions, can do so later if yo'd like